model/map-jpa/src/main/resources/META-INF/groups/jpa-groups-changelog-1.xml

```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright 2022 Red Hat, Inc. and/or its affiliates
+and other contributors as indicated by the @author tags.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd
+                    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+
+    <!-- format of id of changeSet: groups-${org.keycloak.models.map.storage.jpa.Constants.CURRENT_SCHEMA_VERSION_GROUP} -->
+    <changeSet author="keycloak" id="groups-1">
+
+        <createTable tableName="kc_group">
+            <column name="id" type="UUID">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="version" type="INTEGER" defaultValueNumeric="0">
+                <constraints nullable="false"/>
+            </column>
+            <column name="metadata" type="json"/>
+        </createTable>
+        <ext:addGeneratedColumn tableName="kc_group">
+            <ext:column name="entityversion" type="INTEGER" jsonColumn="metadata" jsonProperty="entityVersion"/>
+            <ext:column name="realmid" type="VARCHAR(36)" jsonColumn="metadata" jsonProperty="fRealmId"/>
+            <ext:column name="name" type="VARCHAR(255)" jsonColumn="metadata" jsonProperty="fName"/>
+            <ext:column name="parentid" type="VARCHAR(36)" jsonColumn="metadata" jsonProperty="fParentId"/>
+        </ext:addGeneratedColumn>
+        <createIndex tableName="kc_group" indexName="group_entityVersion">
+            <column name="entityversion"/>
+        </createIndex>
+        <createIndex tableName="kc_group" indexName="group_realmId_name_parentId" unique="true">
+            <column name="realmid"/>
+            <column name="name"/>
+            <column name="parentid"/>
+        </createIndex>
```

Comment:
I'm sorry that I missed that earlier: This index should be IMHO split:

* one unique index to ensure that realmid + name of the group are unique 

* another non-unique index for the parentId so that the foreign key is indexed and simple to look up when traversing the tree upwards. The index for parentId doesn't need to contain the group name and also doesn't need to contain the realmId as parentId is unique across realms ... at least technically, I don't know what query the criteria API constructs here.

Reply:
Thanks for noting why the index needs to have three columns, I stand corrected. I didn't know the same name could appear in separate branches of the group tree. 

To be able to query by parentId, I see the following use cases:

* to validate the foreign key (group is referencing itself) - looking at the columns again, I don't see that foreign key in the schema, there is only a foreign key in the attributes. A foreign key constraint could ensure that the parent exists, and could also delete cascade children.

* to search the leafs of a groups by parent ID (the getSubGroupsStream method as in classic JPA) ... now looking more closely in MapGroupAdapter, this is could be quite inefficient as it streams all groups from the database and then filters :-( Still, one could leave that for a future optimization and create a criteria query in MapGroupAdapter.

Reply:
Thank you for the details. I can now see the potential for the `(realmId, parentId)` index. Could you please create an issue to be addressed re `getSubGroupsStream` inefficiency?

For the sake of comleteness - the validation argument would only hold if the groups were only stored in JPA and only within a single JPA instance - then the index would be required. This assumption does not hold for the tree storage where the parent groups might be stored elsewhere, e.g. in LDAP.